crossdock/client/errorstchout/behavior.go

```diff
@@ -0,0 +1,196 @@
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.Error(err, "is error")
+				assert.False(isAppErr, "no-such-procedure must not produce application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "no-such-procedure must produce  system error")
+				if !ok {
+					return
+				}
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(tchannel.ErrCodeBadRequest, code, "must produce bad request error")
+				assert.Contains(err.Error(), `unrecognized procedure "no-such-procedure"`, "must mention unrecongized procedure in error message")
+			},
+		},
+		{
+			name:      "bad response",
+			procedure: "bad-response",
+			body:      []byte("{}"),
+			headers:   []byte("{}"),
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.Error(err, "is error")
+				assert.False(isAppErr, "bad-response must not produce an application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "bad-response must produce system error")
+				if !ok {
+					return
+				}
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(tchannel.ErrCodeUnexpected, code, "bad-response must produce unexpected error")
+				assert.Contains(err.Error(), `failed to encode "json"`, "must mention failure to encode JSON in error message")
+			},
+		},
+		{
+			name:      "unexpected error",
+			procedure: "unexpected-error",
+			body:      []byte("{}"),
+			headers:   []byte("{}"),
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.Error(err, "is error")
+				assert.False(isAppErr, "unexpected-error procedure must not produce application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "unexpected-error procedure must produce system error")
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(tchannel.ErrCodeUnexpected, code, "must produce transport error")
+			},
+		},
+	}
+
+	server := t.Param(params.Server)
+	serverHostPort := fmt.Sprintf("%v:%v", server, serverPort)
+
+	ch, err := tchannel.NewChannel(serviceName, nil)
+	fatals.NoError(err, "could not create channel")
+
+	peer := ch.Peers().Add(serverHostPort)
+
+	for _, tt := range tests {
+		var res2, res3 []byte
+		var headers map[string]string
+
+		t.Tag("case", tt.name)
+		t.Tag("procedure", tt.procedure)
+
+		ctx, cancel := json.NewContext(time.Second)
+		defer cancel()
+
+		ctx = json.WithHeaders(ctx, headers)
+
+		call, err := peer.BeginCall(
+			ctx,
+			serviceName,
+			tt.procedure,
+			&tchannel.CallOptions{Format: tchannel.Format("json")},
+		)
+		fatals.NoError(err, "could not begin call")
+
+		err = tchannel.NewArgWriter(call.Arg2Writer()).Write(tt.headers)
+		fatals.NoError(err, "could not write request headers")
+
+		err = tchannel.NewArgWriter(call.Arg3Writer()).Write(tt.body)
+		fatals.NoError(err, "could not write request body")
+
+		err = tchannel.NewArgReader(call.Response().Arg2Reader()).Read(&res2)
+		isAppErr := call.Response().ApplicationError()
+		if err != nil {
```

Comment:
Is this right? If you failed to read arg2, you'll overwrite the error with arg3's read error?